src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/PowCoinviewRule.cs

```diff
@@ -1,87 +1,63 @@
 
         /// <summary>Consensus options.</summary>
-        public PowConsensusOptions ConsensusOptions { get; }
+        public PowConsensusOptions ConsensusOptions { get; set; }
 
-        /// <summary>Keeps track of how much time different actions took to execute and how many times they were executed.</summary>
-        public ConsensusPerformanceCounter PerformanceCounter { get; }
-
-        private readonly Network network;
+        /// <inheritdoc />
+        public override void Initialize()
+        {
+            this.Logger.LogTrace("()");
 
-        /// <summary>Provider of time functions.</summary>
-        protected readonly IDateTimeProvider dateTimeProvider;
+            this.ConsensusParams = this.Parent.Network.Consensus;
+            this.ConsensusOptions = this.ConsensusParams.Option<PowConsensusOptions>();
 
-        /// <summary>
-        /// Initializes a new instance of the <see cref="PowConsensusValidator"/> class.
-        /// </summary>
-        /// <param name="network">Specification of the network the node runs on - regtest/testnet/mainnet.</param>
-        /// <param name="checkpoints">Provider of block header hash checkpoints.</param>
-        /// <param name="dateTimeProvider">Provider of time functions.</param>
-        /// <param name="loggerFactory">Factory for creating loggers.</param>
-        public PowConsensusValidator(Network network, ICheckpoints checkpoints, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
-        {
-            Guard.NotNull(network, nameof(network));
-            Guard.NotNull(network.Consensus.Option<PowConsensusOptions>(), nameof(network.Consensus.Options));
-
-            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
-            this.ConsensusParams = network.Consensus;
-            this.ConsensusOptions = network.Consensus.Option<PowConsensusOptions>();
-            this.network = network;
-            this.dateTimeProvider = dateTimeProvider;
-            this.PerformanceCounter = new ConsensusPerformanceCounter(this.dateTimeProvider);
-            this.Checkpoints = checkpoints;
+            this.Logger.LogTrace("(-)");
         }
 
         /// <inheritdoc />
-        public virtual void ExecuteBlock(RuleContext context, TaskScheduler taskScheduler = null)
+        public override Task RunAsync(RuleContext context)
         {
-            this.logger.LogTrace("()");
+            this.Logger.LogTrace("()");
 
             Block block = context.BlockValidationContext.Block;
             ChainedHeader index = context.BlockValidationContext.ChainedHeader;
             DeploymentFlags flags = context.Flags;
             UnspentOutputSet view = context.Set;
 
-            this.PerformanceCounter.AddProcessedBlocks(1);
-            taskScheduler = taskScheduler ?? TaskScheduler.Default;
+            this.Parent.PerformanceCounter.AddProcessedBlocks(1);
+            var taskScheduler = TaskScheduler.Default;
```

Comment:
This variable is not needed. 

`checkInput.Start(taskScheduler);` : 118 - this can be called without any params- default Task Scheduler will be used

Reply:
ah cool